Pagination, sorting, swagger

ponomvrenko opened the pull request:
Configuration:
- pom.xml: added <springdoc-openapi-starter-webmvc-ui> dependency

DTO:
- CreateBookRequestDto: added @Size to author field.

Service:
- BookServiceImpl & BookService: Implement pagination(sort) logic. Added Pageable to as a parameter in getAll() method.

Controller:
- BookController: Added annotations to document my code. (swagger)

ponomvrenko committed:
Configuration:
- pom.xml: added <springdoc-openapi-starter-webmvc-ui> dependency

DTO:
- CreateBookRequestDto: added @Size to author field.

Service:
- BookServiceImpl & BookService: Implement pagination(sort) logic. Added Pageable to as a parameter in getAll() method.

Controller:
- BookController: Added annotations to document my code. (swagger)

RostyslavKuzmych commented on pom.xml:
```diff
@@ -80,6 +80,15 @@
+            <artifactId>spring-boot-starter-validation</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.springdoc</groupId>
```
I guess, that we work with swagger, not swager.

RostyslavKuzmych commented on src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java:
```diff
@@ -1,13 +1,24 @@
+    @Size(min = 10, max = 13)
     private String isbn;
+    @NotNull
+    @Min(0)
```
We can use @Positive

RostyslavKuzmych commented on src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -0,0 +1,55 @@
+    @Override
+    protected ResponseEntity<Object> handleMethodArgumentNotValid(
+            MethodArgumentNotValidException ex,
+            @NonNull HttpHeaders headers,
+            @NonNull HttpStatusCode status,
+            @NonNull WebRequest request) {
```
I guess, that is redundant

RostyslavKuzmych commented on src/main/java/com/example/onlinebookstore/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    @ExceptionHandler(EntityNotFoundException.class)
+    public ResponseEntity<Object> handleEntityNotFoundException(Exception ex) {
```
I guess, that we have to receive EntityNotFoundException, not Exception

RostyslavKuzmych commented on src/main/java/com/example/onlinebookstore/service/impl/BookServiceImpl.java:
```diff
@@ -45,9 +48,15 @@ public void deleteById(Long id) {
-        Book book = bookMapper.toModel(requestDto);
-        book.setId(id);
-        return bookMapper.toDto(bookRepository.save(book));
+    public BookDto updateById(Long id, CreateBookRequestDto requestDto) {
+        try {
+            Book book = bookMapper.toModel(requestDto);
+            book.setId(id);
+            return bookMapper.toDto(bookRepository.save(book));
+        } catch (EntityNotFoundException e) {
+            throw new EntityNotFoundException(
+                    "The book with ID: id " + id + ", unfortunately was not be founded."
+            );
```
I doubt, if we must use that, and think that it's redundant too, and i just wonder, what will be, if input id will be 100000, and we have only 100 books in the db, it turns out that we will have a disorder in the db...

ponomvrenko replied:
Hey @RostyslavKuzmych, thanks for the comment, but I can’t get the main idea of it. 
We have a method that updates our book by ID, also there is a check that validates our ID. 
Question is, what’s the redundant thing?

bhdnchui replied:
There should be no disorder in the database, but this is provided by another functionality.
Regarding this method, here you need to check if there is a book with the received id, and only if it is - update it, otherwise, throw an error. Rewrite.

RostyslavKuzmych reviewed (changes requested):
Overall good, but pay attention to my comments, please.

ponomvrenko committed:
Small fixes after student review.

bhdnchui commented on src/main/java/com/example/onlinebookstore/controller/BookController.java:
```diff
@@ -24,28 +29,39 @@ public class BookController {
     @DeleteMapping("/{id}")
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @Operation(summary = "Delete book by ID",
+            description = "\"Softly\" removes the book")
```
I don't think that the user must know that delete is soft. 
It is our implementation and user shouldn't know that, in otherwise, he will start asking questions about how to delete the book forever.

bhdnchui commented on src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java:
```diff
@@ -1,13 +1,24 @@
+    @Size(min = 1, max = 40)
     private String author;
+    @NotNull
+    @Size(min = 10, max = 13)
```
Think about using org.hibernate.validator.constraints.ISBN annotation for validation.

bhdnchui commented on src/main/java/com/example/onlinebookstore/dto/CreateBookRequestDto.java:
```diff
@@ -1,13 +1,24 @@
 
 @Data
 public class CreateBookRequestDto {
+    @NotNull
```
Can the title be an empty string?

bhdnchui reviewed (changes requested):
Good job! Just a few points to consider.

ponomvrenko committed:
Optimized description in deleteBookById;
On CreateBookRequestDto used @ISBN annotation for isbn field instead of @Size;
Added at BookServiceImpl checking for existing book in repository (updateById method).

d1sam reviewed (changes requested):
Why do I see changes, that not related to this homework? There are some changes that related to hw GlobalExceptionHandler and data validation. Please fix it :)

ponomvrenko committed:
Merge branch 'master' of https://github.com/ponomvrenko/online-book-store into pagination-sorting-swager

vovasalyha reviewed (approved):
LGTM
